src/log_batch.rs

```diff
@@ -392,208 +332,363 @@ impl<M: MessageExt> LogItem<M> {
             content,
         })
     }
+
+    fn compute_approximate_size(&self) -> usize {
+        match &self.content {
+            LogItemContent::EntriesIndex(entries_index) => {
+                8 /*r_id*/ + 1 /*type*/ + entries_index.compute_approximate_size()
+            }
+            LogItemContent::Command(cmd) => 8 + 1 + cmd.compute_approximate_size(),
+            LogItemContent::Kv(kv) => 8 + 1 + kv.compute_approximate_size(),
+        }
+    }
 }
 
 #[derive(Debug, PartialEq)]
-pub struct LogBatch<M: MessageExt> {
-    items: Vec<LogItem<M>>,
-    items_size: usize,
+pub struct LogItemBatch<M: MessageExt> {
+    items: Vec<LogItem>,
     _phantom: PhantomData<M>,
 }
 
-impl<M> Default for LogBatch<M>
+impl<M> Default for LogItemBatch<M>
 where
     M: MessageExt,
 {
     fn default() -> Self {
-        Self {
-            items: Vec::with_capacity(DEFAULT_BATCH_CAP),
-            items_size: 0,
-            _phantom: PhantomData,
-        }
+        Self::with_capacity(DEFAULT_BATCH_CAP)
     }
 }
 
-impl<M> LogBatch<M>
+impl<M> LogItemBatch<M>
 where
     M: MessageExt,
 {
     pub fn with_capacity(cap: usize) -> Self {
         Self {
             items: Vec::with_capacity(cap),
-            items_size: 0,
             _phantom: PhantomData,
         }
     }
 
-    pub fn merge(&mut self, rhs: Self) {
-        self.items_size += rhs.items_size;
-        self.items.extend(rhs.items);
+    pub fn merge(&mut self, rhs: &mut Self) {
+        self.items.append(&mut rhs.items);
+    }
+
+    pub fn iter_mut(&mut self) -> core::slice::IterMut<LogItem> {
+        self.items.iter_mut()
     }
 
-    pub fn drain(&mut self) -> std::vec::Drain<'_, LogItem<M>> {
-        self.items_size = 0;
+    pub fn drain(&mut self) -> std::vec::Drain<'_, LogItem> {
         self.items.drain(..)
     }
 
+    pub fn is_empty(&self) -> bool {
+        self.items.is_empty()
+    }
```

Comment:
I think these can be removed, we are implementing them for `LogBatch` to hide data members from other modules.